### PR TITLE
[MRESOLVER-346] Too eager locking in resolver

### PR DESCRIPTION
**Pull request by cstamas**
The locking in resolver is too eager. This PR relaxes locking by introducing use of shared locks (not used before) and following changes:
* installer -- no change, remains exclusive locking
* deployer -- laxed to shared locking, as it only reads local repo
* Artifact and Metadata resolver -- implemented "upgrade", they are optimistic and start with shared lock (and will happily finish if local repo has all) but will "upgrade" to exclusive if remote access (hence local caching) is about to happen.

Best viewed with whitespace ignore, as there are not much change but blocks got nested.

---

https://issues.apache.org/jira/browse/MRESOLVER-346

### CODE REVIEW

**Comment by michael-o** [comment]
I will review this next week

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/synccontext/named/NamedLockFactoryAdapter.java [hunk 1]
```diff
@@ -48,6 +49,14 @@ public final class NamedLockFactoryAdapter {
 
     public static final TimeUnit DEFAULT_TIME_UNIT = TimeUnit.SECONDS;
 
+    public static final String RETRIES_KEY = "aether.syncContext.named.retries";
```
Comment (slawekjaranowski) [inline maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/synccontext/named/NamedLockFactoryAdapter.java:52]: New options should be add to documentation.

Comment (cstamas) [reply maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/synccontext/named/NamedLockFactoryAdapter.java:52]: Sure, but pls consider this PR as "under testing", this is what users on ML tested (I provided them "unofficial builds" via Google Drive): https://lists.apache.org/thread/q4nlojworr8lynroyv2j0721vpmpgh70

Will add doco next.

Comment (cstamas) [reply maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/synccontext/named/NamedLockFactoryAdapter.java:52]: Doco added

**Comment by michael-o** [comment]
> The locking in resolver is too eager. This PR relaxes locking by introducing use of shared locks (not used before) and following changes:
> 
>     * installer -- no change, remains exclusive locking
> 
>     * deployer -- laxed to shared locking, as it only reads local repo
> 
>     * A and M resolver -- implemented "upgrade", they are optimistic and start with shared lock (and will happily finish if local repo has all) but will "upgrade" to exclusive if remote access (hence local caching) is about to happen.
> 
>     * syncContext adapter -- do not "hold" on acquired locks if one fails, but release all/retry all
> 
> 
> Best viewed with whitespace ignore, as there are not much change but blocks got nested.
> 
> https://issues.apache.org/jira/browse/MRESOLVER-346

So far, you explanation sounds reasonable to me. Reviewing...

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/DefaultArtifactResolver.java [hunk 2]
```diff
@@ -256,207 +257,226 @@ public List<ArtifactResult> resolveArtifacts(
...
-                        artifact,
-                        remoteRepositories);
-            }
-
-            LOGGER.debug("Resolving artifact {} from {}", artifact, remoteRepositories);
-            AtomicBoolean resolved = new AtomicBoolean(false);
-            Iterator<ResolutionGroup> groupIt = groups.iterator();
-            for (RemoteRepository repo : filteredRemoteRepositories) {
-                if (!repo.getPolicy(artifact.isSnapshot()).isEnabled()) {
+                if (!groups.isEmpty() && current == shared) {
```
Comment (michael-o) [inline maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/DefaultArtifactResolver.java:441]: The not empty implies that resolution is not complete?

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/DefaultArtifactResolver.java [hunk 3]
```diff
@@ -256,207 +257,226 @@ public List<ArtifactResult> resolveArtifacts(
...
+            Collection<Artifact> subjects,
+            RepositorySystemSession session,
+            Collection<? extends ArtifactRequest> requests)
             throws ArtifactResolutionException {
-        List<ArtifactResult> results = new ArrayList<>(requests.size());
-        boolean failures = false;
-        final boolean simpleLrmInterop = ConfigUtils.getBoolean(session, false, CONFIG_PROP_SIMPLE_LRM_INTEROP);
+        SyncContext current = shared;
+        try {
+            while (true) {
```
Comment (michael-o) [inline maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/DefaultArtifactResolver.java:274]: Can we end up in an endless loop here? Whould a do while loop with a condition make more sense here? We need at least one iteration.

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/DefaultDeployer.java [hunk 4]
```diff
@@ -197,7 +197,7 @@ public DeployResult deploy(RepositorySystemSession session, DeployRequest reques
                     e);
         }
 
-        try (SyncContext syncContext = syncContextFactory.newInstance(session, false)) {
+        try (SyncContext syncContext = syncContextFactory.newInstance(session, true)) {
```
Comment (michael-o) [inline maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/DefaultDeployer.java:200]: Checksums and stuff are generated on the fly and never persisted on disk, as far as I remember?!

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/synccontext/named/NamedLockFactoryAdapter.java (hunk 1 above)
Comment (michael-o) [inline maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/synccontext/named/NamedLockFactoryAdapter.java:52]: This change is for me is not related and shoud be a separate ticket and PR.

**Review by michael-o — CHANGES_REQUESTED** [review]

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/DefaultDeployer.java (hunk 4 above)
Comment (cstamas) [reply maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/DefaultDeployer.java:200]: yes, local repo is NOT modified, everything happens on the fly and only remote repo gets content changes

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/DefaultArtifactResolver.java (hunk 2 above)
Comment (cstamas) [reply maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/DefaultArtifactResolver.java:441]: Yes, groups are "prepared download batches" (per remote repo, 1 or more artifact). By group being not empty, means that resolver could not fulfill resolution from local repo, hence is about to download. Here, the "downside" is that when this happens, and we switch from shared to exclusive lock, groups are rebuilt from scratch.

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/DefaultArtifactResolver.java (hunk 3 above)
Comment (cstamas) [reply maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/DefaultArtifactResolver.java:274]: I don't see how: last element within this while-true loop is `return results`, meaning if we reach here even only once, we are out of the loop.

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/synccontext/named/NamedLockFactoryAdapter.java (hunk 1 above)
Comment (cstamas) [reply maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/synccontext/named/NamedLockFactoryAdapter.java:52]: I argue here: this all PR is somewhat about some sort of "backoff" for locking, so IMO this change is also essential part of whole.

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/DefaultArtifactResolver.java (hunk 2 above)
Comment (michael-o) [reply maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/DefaultArtifactResolver.java:441]: Makes sense!

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/synccontext/named/NamedLockFactoryAdapter.java (hunk 1 above)
Comment (michael-o) [reply maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/synccontext/named/NamedLockFactoryAdapter.java:52]: I still don't share this opinion. While I don't object this feature as a whole, it should be visiable as as separate issue in release notes. Let's focus on too eager locking first...

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/DefaultArtifactResolver.java (hunk 3 above)
Comment (michael-o) [reply maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/DefaultArtifactResolver.java:274]: Right.

**Review by michael-o — APPROVED** [review]
This looks very reasonable now to me.

**Comment by jira-importer** [comment]
Resolve #1023

**Pull request merged**